### ISSUE TITLE
[IMPROVEMENT] Introduce CPU and Memory Requests/Limits for Share Manager Pods via StorageClass Parameters in Longhorn

### ISSUE THREAD

**Issue by derekbit** [longhorn/longhorn#8314]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

At present, Longhorn lacks the functionality to set CPU and memory requests/limits for share manager pods of RWX volumes, leading to unregulated resource consumption in environments sensitive to resource utilization.

To help users better manage the resources consumed by Share Manager pods, we can introduce new parameters including cpu/memory requests/limits in the `storageClass.parameters`. These parameters would specify resource requests/limits that would apply to the corresponding share manager pods.

**Comment by derekbit** [comment]
Ref: https://github.com/longhorn/longhorn/issues/6351

**Comment by derekbit** [comment]
During the hackweek, I just learned about the MemoryQoS concept in Kubernetes, which classifies pods into three categories: `Guaranteed`, `Burstable`, and `Best-Effort`. Here's how they differ:

- Guaranteed: Both memory requests and limits are set to the same value.
- Burstable: Only memory requests are set.
- Best-Effort: Neither memory requests nor limits are set.

In cases of memory pressure on the host, the kubelet prioritizes killing pods in the following order: Best-Effort pods are terminated first, followed by Burstable, and Guaranteed pods are terminated last.

The share-manager pod is a critical component for a RWX volume. However, it is currently classified as a `Best-Effort` pod because its memory request is not set.

We should at least configure the memory request for the share-manager pod, like the instance-manager pod. This change can ensure that the share-manager pod is classified as Burstable, making it not the first to be deleted during memory pressure on the host.

Ref: https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/


WDYT @PhanLe1010 @shuo-wu @james-munson @innobead

**Comment by james-munson** [comment]
I agree - putting a memory size request on the share-manager pod and making it Burstable would be a good idea.  It wouldn't have to be very large, and probably shouldn't be, to avoid problems when there are a lot of RWX volumes.  Something like 50 MiB, perhaps.
We should avoid putting on a limit, though.

**Comment by innobead** [comment]
> The share-manager pod is a critical component for a RWX volume. However, it is currently classified as a Best-Effort pod because its memory request is not set.

Sounds good, as the concept is similar to what we set requests to IM pods. 

> Burstable: Only memory requests are set.

In the best-effort case, Do we know how the pods with memory requests are being selected? Does it depend on the size of the requested memory?

**Comment by derekbit** [comment]
> In the best-effort case, Do we know how the pods with memory requests are being selected? Does it depend on the size of the requested memory?

Do you mean `best-effort` or `burstable`?

**Comment by derekbit** [comment]
In burstable mode, the requested memory is guaranteed and not reclaimable by the Linux kernel. If a memory limit is not set, the pod can consume additional memory as needed, but the additional cache usage can be reclaimed by the kernel.

So, as @james-munson mentioned in https://github.com/longhorn/longhorn/issues/8314#issuecomment-2496289582, we can just assign a small number, e.g. 50 MiB (according to my test, the nfs-ganesha process at least uses 30-40 MiB), then the pod can be set to burstable mode.

**Comment by PhanLe1010** [comment]
> We should at least configure the memory request for the share-manager pod, like the instance-manager pod. This change can ensure that the share-manager pod is classified as Burstable, making it not the first to be deleted during memory pressure on the host.

I agree. This and with priorityclass setting would make share-manager pod to have higher chance of survival

**Comment by derekbit** [comment]
cc @PhanLe1010
